In [56]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [57]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [58]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [59]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [60]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data)

In [61]:
# Create a DataFrame with the scaled data
scaled_marketdf = pd.DataFrame(crypto_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

# Copy the crypto names from the original data
scaled_marketdf['coin_id']=list(df_market_data.index)

# Set the coinid column as index
scaled_marketdf = scaled_marketdf.set_index('coin_id')

# Display sample data
scaled_marketdf.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [62]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

---

### Find the Best Value for k Using the Original Data.

In [63]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [64]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled` (mine is scaled_marketdf)
# 3. Append the model.inertia_ to the inertia list

for i in k: 
    modelk = KMeans(n_clusters=i, random_state=1)
    modelk.fit(scaled_marketdf)
    inertia.append(modelk.inertia_)

inertia

[287.0,
 195.82021818036043,
 123.19048183836958,
 79.02243535120975,
 65.40592346140596,
 52.933558921015006,
 47.98312409811001,
 37.28818726271726,
 33.06168486647882,
 28.779752431429223]

In [65]:
# Create a dictionary with the data to plot the Elbow curve
elbow = {'k':k, 'inertia':inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df1 = pd.DataFrame(elbow)
elbow_df1

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.405923
5   6   52.933559
6   7   47.983124
7   8   37.288187
8   9   33.061685
9  10   28.779752

In [66]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df1.hvplot.line(x='k', y='inertia', xticks=k)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [67]:
# Initialize the K-Means model using the best value for k (=4)
kmodel = KMeans(n_clusters=4, random_state=1)

In [68]:
# Fit the K-Means model using the scaled data
kmodel.fit(scaled_marketdf)

KMeans(n_clusters=4, random_state=1)

In [69]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_predictions = kmodel.predict(scaled_marketdf)

# Print the resulting array of cluster values.
k_predictions

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1],
      dtype=int32)

In [70]:
# Create a copy of the DataFrame
scaled_df_predictions = scaled_marketdf.copy()

In [71]:
# Add a new column to the DataFrame with the predicted clusters
scaled_df_predictions['k_predictions'] = k_predictions

# Display sample data
scaled_df_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  k_predictions  
coin_id                                                  
bitcoin                        -0.251637              3  
ethereum                       -0.199352              3  
tether                         -0.282061              1  
ripple                         -0.295546              1  
bitcoin-cash                   -0.270317              3

In [72]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

scaled_df_predictions.hvplot.scatter(x='price_change_percentage_24h', y='price_change_percentage_7d', by='k_predictions', hover_cols= 'coin_id')

:NdOverlay   [k_predictions]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [73]:
# Create a PCA model instance and set `n_components=3`.
pca= PCA(n_components=3)

In [74]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_fit = pca.fit_transform(scaled_df_predictions)
# View the first five rows of the DataFrame. 
pca_fit[0:5]

array([[-1.58513942,  0.13439444,  0.545621  ],
       [-1.32258066,  0.05449393,  1.11570854],
       [ 0.26633536, -0.53640937, -0.79694588],
       [ 0.25722513, -0.59201286, -0.6465287 ],
       [-2.95224985,  0.42237155,  1.26789353]])

In [75]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.37005408, 0.32322221, 0.19115222])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?
.37+.323+.191 = 
**Answer:** .884

In [76]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(pca_fit, columns=['PCA1', 'PCA2', 'PCA3'])
# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data
pca_df['coin_id'] = list(scaled_df_predictions.index)

# Set the coinid column as index
pca_df = pca_df.set_index('coin_id')

# Display sample data
pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -1.585139  0.134394  0.545621
ethereum     -1.322581  0.054494  1.115709
tether        0.266335 -0.536409 -0.796946
ripple        0.257225 -0.592013 -0.646529
bitcoin-cash -2.952250  0.422372  1.267894

---

### Find the Best Value for k Using the PCA Data

In [77]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [78]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca` - mine is pca_df
# 3. Append the model.inertia_ to the inertia list
for i in k: 
    modelk = KMeans(n_clusters=i, random_state=1)
    modelk.fit(pca_df)
    inertia.append(modelk.inertia_)

inertia

[287.00783759231985,
 196.26835760962865,
 111.97202664282966,
 43.205878719552985,
 33.042435121783235,
 24.347640212866626,
 19.77731971115921,
 15.868896599933997,
 12.35105773925925,
 10.16935089732053]

In [79]:
# Create a dictionary with the data to plot the Elbow curve
elbow = {'k':k, 'inertia':inertia}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df2 = pd.DataFrame(elbow)

In [80]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df2.hvplot.line(x='k', y='inertia', xticks=k)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Answer is still 4, but it seems more obvious here. 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [81]:
# Initialize the K-Means model using the best value for k
kmeans_model = KMeans(n_clusters=3, random_state=1)

In [82]:
# Fit the K-Means model using the PCA data
kmeans_model.fit(pca_df)

KMeans(n_clusters=3, random_state=1)

In [83]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k3_predictions = kmeans_model.predict(pca_df)
# Print the resulting array of cluster values.
k3_predictions

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 1, 0, 0, 0, 0],
      dtype=int32)

In [84]:
# Create a copy of the DataFrame with the PCA data
k_predictions = pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
k_predictions['k_predictions'] = k3_predictions

# Display sample data
k_predictions

PCA1      PCA2      PCA3  k_predictions
coin_id                                                          
bitcoin              -1.585139  0.134394  0.545621              1
ethereum             -1.322581  0.054494  1.115709              1
tether                0.266335 -0.536409 -0.796946              0
ripple                0.257225 -0.592013 -0.646529              0
bitcoin-cash         -2.952250  0.422372  1.267894              1
binancecoin          -1.975484  0.528219  0.670145              1
chainlink            -1.671906  0.178423  2.718598              1
cardano              -1.539740  0.324517  1.497226              1
litecoin             -1.347533 -0.132837  0.791063              1
bitcoin-cash-sv       0.040698 -0.802083 -0.465562              0
crypto-com-chain      1.344822 -1.108679 -1.076912              0
usd-coin              0.272121 -0.545849 -0.814012              0
eos                   0.320673 -0.918019 -0.637918              0
monero               -2.838740  1.895142  0.848863              1
tron                  0.143758 -0.442297 -0.274682              0
tezos                -0.862155 -0.757488  1.490647              1
okb                   1.378653 -0.776673 -0.754383              0
stellar               0.532658 -0.876910 -0.120776              0
cosmos               -1.539433  0.352390  1.713933              1
cdai                  0.208186 -0.588555 -0.830587              0
neo                   0.887872 -0.947174 -0.588142              0
wrapped-bitcoin      -1.572600  0.121709  0.530894              1
leo-token             0.697532 -0.826190 -0.993080              0
huobi-token           0.263251 -0.527831 -0.530039              0
nem                   0.408806  0.707042 -1.330139              0
binance-usd           0.246128 -0.540957 -0.809982              0
iota                  0.186495 -0.967371  0.003609              0
vechain               1.528493 -1.668876 -0.750235              0
zcash                -1.347135  0.106093  1.200966              1
theta-token           2.056222  2.034190 -1.525531              0
dash                  0.222043 -0.817068  0.122347              0
ethereum-classic      0.228151 -0.739212 -0.299732              0
ethlend               7.458259  4.338598  3.751968              2
maker                -0.991935 -0.079763  0.741439              1
havven                2.271195 -0.622938  0.875803              0
omisego               0.234856  0.223448 -1.390103              0
celsius-degree-token -1.949062  7.705390 -3.110327              1
ontology              1.510753 -1.801827 -0.193429              0
ftx-token            -0.092618 -0.521653 -0.153336              0
true-usd              0.229197 -0.538204 -0.805785              0
digibyte              0.393926 -0.449545 -0.988556              0

In [85]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
k_predictions.hvplot.scatter(x='PCA1', y='PCA2', by='k_predictions', hover_cols='coin_id')

:NdOverlay   [k_predictions]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [88]:
# Composite plot to contrast the Elbow curves
elbow_df1.hvplot.line(x='k', y='inertia', xticks=k, title='Elbow Curve 1: KMeans') + elbow_df2.hvplot.line(x='k', y='inertia', xticks=k, title='Elbow Curve 2: PCA')

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [89]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
scaled_df_predictions.hvplot.scatter(x='price_change_percentage_24h', y='price_change_percentage_7d', by='k_predictions', hover_cols= 'coin_id', title ='KMeans Scatter Plot') + k_predictions.hvplot.scatter(x='PCA1', y='PCA2', by='k_predictions', hover_cols='coin_id', title='PCA Scatter Plot')

:Layout
   .NdOverlay.I  :NdOverlay   [k_predictions]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [k_predictions]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** There is a clearer distinction between the clusters with PCA1 and PCA2. The data is closer together, whereas there's more spread with k_predictions. On the elbow curve we also see a much clearer option for the numebr of clusters to use on Kmeans. 